### ISSUE TITLE
About injective mapping and NCE loss in Instance Loss part.

### ISSUE THREAD

**Issue by TIMESTICKING** [nihalsid/panoptic-lifting#11]
Hi, and thx for such wonderful work! After reading ur articles, I remain a few questions and hope that u would open a discussion panel, but for now, I just put my questions here :)

1. What is the form of the injective mapping function and which part of the code is related to it
2. Is the mapping function related to the training image? Like there's one mapping function for one image?
3. Where or what is the NCE loss, since I found `RCE` in ur code

Anyway, I'm just very confused about the Instance Loss part. I'd really appreciate it if you guys could explain it more. Thank you!

**Comment by nihalsid** [comment]
Hi! Thank you for your interest in our work :) I'll try to answer your questions here:

1. [This](https://github.com/nihalsid/panoptic-lifting/blob/7af7a3e8477ead8e57f699a240d993e3bc21ee42/trainer/train_panopli_tensorf.py#L189) line calls the function that handles mapping 2D mask2former labels (labels_gt) to the labels used to supervise the surrogate IDs (virtual_gt_labels). 

2. This mapping is per batch of rays, and this batch of rays is sampled from a single image. I say "per batch" because the mapping is can possibly be changing per training iteration, since it depends on the current prediction scores.

3. We use the pytorch cross entropy loss ([here](https://github.com/nihalsid/panoptic-lifting/blob/7af7a3e8477ead8e57f699a240d993e3bc21ee42/trainer/train_panopli_tensorf.py#L64)). The RCE is something I tried but didn't see any improvement, and is not used.

Hope this helps!

**Comment by nihalsid** [comment]
Hi,

For the segment clustering loss, at some point I rewrote it to be calculate with torch_scatter for performance reasons: https://github.com/nihalsid/panoptic-lifting/blob/7af7a3e8477ead8e57f699a240d993e3bc21ee42/trainer/train_panopli_tensorf.py#L162

The `calculate_segment_clustering_loss()` was an old inefficient implementation and is no longer used.

Let me look at the equation and get back to you about it in a separate message.

**Comment by nihalsid** [comment]
Yes, you're right, the shortened equation you write makes sense, but that would be the `prob_conf` version instead of  `argmax_conf` in the `calculate_segment_clustering_loss`. We ended up using the `argmax_conf` version where the cross entropy is calculated against the one hot label instead of the mean probability distribution of the group.

Hope it helps!
